StenosisMeasurement2D/StenosisMeasurement2D.py

```diff
@@ -0,0 +1,750 @@
+import os
+import unittest
+import logging
+import vtk, qt, ctk, slicer
+from slicer.ScriptedLoadableModule import *
+from slicer.util import VTKObservationMixin
+
+#
+# StenosisMeasurement2D
+#
+
+class StenosisMeasurement2D(ScriptedLoadableModule):
+  """Uses ScriptedLoadableModule base class, available at:
+  https://github.com/Slicer/Slicer/blob/master/Base/Python/slicer/ScriptedLoadableModule.py
+  """
+
+  def __init__(self, parent):
+    ScriptedLoadableModule.__init__(self, parent)
+    self.parent.title = "Stenosis measurement : 2D"
+    self.parent.categories = ["Vascular Modeling Toolkit"]
+    self.parent.dependencies = []
+    self.parent.contributors = ["SET [Surgeon] [Hobbyist developer]", "Andras Lasso, PerkLab"]
+    self.parent.helpText = """
+This <a href="https://github.com/vmtk/SlicerExtension-VMTK/">module</a> calculates the surface area of segments cut by a slice plane in its orientation. It is intended for quick two dimensional arterial stenosis evaluation, but is actually purpose agnostic.
+"""
+    # TODO: replace with organization, grant and thanks
+    self.parent.acknowledgementText = """
+This file was originally developed by Jean-Christophe Fillion-Robin, Kitware Inc., Andras Lasso, PerkLab,
+and Steve Pieper, Isomics, Inc. and was partially funded by NIH grant 3P41RR013218-12S1.
+"""
+
+class StenosisMeasurement2DWidget(ScriptedLoadableModuleWidget, VTKObservationMixin):
+  """Uses ScriptedLoadableModuleWidget base class, available at:
+  https://github.com/Slicer/Slicer/blob/master/Base/Python/slicer/ScriptedLoadableModule.py
+  """
+
+  def __init__(self, parent=None):
+    """
+    Called when the user opens the module the first time and the widget is initialized.
+    """
+    ScriptedLoadableModuleWidget.__init__(self, parent)
+    VTKObservationMixin.__init__(self)  # needed for parameter node observation
+    self.logic = None
+    self._parameterNode = None
+    self._updatingGUIFromParameterNode = False
+    """
+    N.B. - We store slice orientation in NthControlPointOrientationMatrixWorld.
+    Unfortunately, this data is not saved in the markups JSON file,
+    though there is an orientation key.
+    """
+    self.currentFiducialNode = None
+    self.currentControlPointID = "-1" # API gets/sets it as a string
+    self.optionResetControlPointOrientationMatrix = False # Checkable menu item
+    self.optionCreateCutModel = False # Checkable menu item
+    """
+    We are using JumpToPointEvent of the fiducial display node.
+    It gets fired 3 times when we click on a control point.
```

Comment:
It is fired only once. You can confirm this using this code snippet:

```python
fdn=getNode('F').GetDisplayNode()
observation = fdn.AddObserver(slicer.vtkMRMLMarkupsFiducialDisplayNode.JumpToPointEvent, lambda a, b: print("JumpToPointEvent!"))
```

If you see the event called multiple times then most likely you add the observer multiple times.

Reply:
Fixed, it was fired in multiple places in the module's code.